kernel/arch/dreamcast/include/dc/vec3f.h

```diff
@@ -37,408 +41,227 @@ typedef struct vec3f {
-    \param  x2               The X coordinate of second vector.
-    \param  y2               The Y coordinate of second vector.
-    \param  z2               The Z coordinate of second vector.
-    \param  w                The result of the calculation.
+    \param  vec1             The first vector.
+    \param  vec2             The second vector.
+    \return                  The result of the calculation.
 */
-#define vec3f_distance(x1, y1, z1, x2, y2, z2, w) { \
-        register float __x  __asm__("fr0") = (x2-x1); \
-        register float __y  __asm__("fr1") = (y2-y1); \
-        register float __z  __asm__("fr2") = (z2-z1); \
-        register float __w  __asm__("fr3"); \
-        __asm__ __volatile__( \
-                              "fldi0 fr3\n" \
-                              "fipr  fv0,fv0\n" \
-                              "fsqrt fr3\n" \
-                              : "+f" (__w) \
-                              : "f" (__x), "f" (__y), "f" (__z), "f" (__w) \
-                            ); \
-        w = __w; \
-    }
-
-/** \brief  Macro to return the normalized version of a vector.
+static inline float vec_distance(vec3f_t vec1, vec3f_t vec2) {
+    vec3f_t vec = (vec3f_t){ vec2.x - vec1.x, vec2.y - vec1.y, vec2.z - vec1.z };
+    return vec_length(vec);
+}
+
+/** \brief  Function to return the normalized version of a vector.
 
     This macro is an inline assembly operation using the SH4's fast
     (approximate) math instructions to calculate a vector that is in the same
-    direction as the input vector but with a Euclidean length of one. The input
-    vector is modified by the operation as the resulting values.
+    direction as the input vector but with a Euclidean length of one.
 
-    \param  x               The X coordinate of vector.
-    \param  y               The Y coordinate of vector.
-    \param  z               The Z coordinate of vector.
+    \param  vec             The source vector.
+    \return                 The normalized vector.
 */
-#define vec3f_normalize(x, y, z) { \
-        register float __x __asm__("fr0") = x; \
-        register float __y __asm__("fr1") = y; \
-        register float __z __asm__("fr2") = z; \
-        __asm__ __volatile__( \
-                              "fldi0 fr3\n" \
-                              "fipr  fv0,fv0\n" \
-                              "fsrra fr3\n" \
-                              "fmul  fr3, fr0\n" \
-                              "fmul  fr3, fr1\n" \
-                              "fmul  fr3, fr2\n" \
-                              : "=f" (__x), "=f" (__y), "=f" (__z) \
-                              : "0" (__x), "1" (__y), "2" (__z) \
-                              : "fr3" ); \
-        x = __x; y = __y; z = __z; \
-    }
-
-/** \brief  Macro to return the normalized version of a vector minus another
+static inline vec3f_t vec_normalize(vec3f_t vec) {
+    float factor = 1.0f / vec_length(vec);
```

Comment:
```suggestion
    float length = vec_length(vec);
    float factor = 1.0f / sqrtf(length * length);
```
If we want this shit to be as fast as possible, shouldn't we be using the fast division approximations?

Reply:
Your suggestion is *not* equivalent.
And with the current code it is already using `fsrra`.

Reply:
Ok, reading it again I guess it's equivalent.
However, what you're missing is that `vec_length()` already contains a `sqrtf()`, so the whole thing is converted to a FSRRA already.

Reply:
Ahhhh! Yeah, you're right, I guess it should be smart enough to see the `sqrtf()` through the inline call, huh...